File: go/verify_solutions.go
package main

import (
	"bytes"
	"flag"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

// ANSI color codes
const (
	colorReset  = "\033[0m"
	colorGreen  = "\033[92m"
	colorRed    = "\033[91m"
	colorYellow = "\033[93m"
	colorBlue   = "\033[94m"
)

// getTimeEmoji returns an emoji based on execution time
func getTimeEmoji(elapsedTime float64) string {
	switch {
	case elapsedTime < 1.0:
		return "⚡"
	case elapsedTime < 3.0:
		return "🚀"
	case elapsedTime < 10.0:
		return "▶️"
	case elapsedTime < 30.0:
		return "🐢"
	default:
		return "🐌"
	}
}

// getExpectedAnswer reads the expected answer from aoc-data directory
func getExpectedAnswer(year, day, part int) (string, bool) {
	answerFile := filepath.Join("..", "aoc-data", strconv.Itoa(year), strconv.Itoa(day), fmt.Sprintf("solution-%d", part))

	data, err := os.ReadFile(answerFile)
	if err != nil {
		return "", false
	}

	content := strings.TrimSpace(string(data))
	// Return false if empty or just '0'
	if content == "" || content == "0" {
		return "", false
	}

	return content, true
}

// writeAnswer writes an answer to the aoc-data directory
func writeAnswer(year, day, part int, answer string) bool {
	answerDir := filepath.Join("..", "aoc-data", strconv.Itoa(year), strconv.Itoa(day))
	answerFile := filepath.Join(answerDir, fmt.Sprintf("solution-%d", part))

	// Create directory if it doesn't exist
	if err := os.MkdirAll(answerDir, 0755); err != nil {
		fmt.Printf("%sError creating directory: %v%s\n", colorRed, err, colorReset)
		return false
	}

	// Write the answer
	if err := os.WriteFile(answerFile, []byte(strings.TrimSpace(answer)+"\n"), 0644); err != nil {
		fmt.Printf("%sError writing answer: %v%s\n", colorRed, err, colorReset)
		return false
	}

	return true
}

// runSolution runs a Go solution and extracts the answer
func runSolution(year, day, part int) (string, bool, float64, bool) {
	dayStr := fmt.Sprintf("%02d", day)
	solutionFile := filepath.Join(strconv.Itoa(year), dayStr, fmt.Sprintf("solution%d.go", part))

	// Check if file exists
	if _, err := os.Stat(solutionFile); os.IsNotExist(err) {
		return "", false, 0.0, false
	}

	// Run the solution
	startTime := time.Now()
	cmd := exec.Command("go", "run", solutionFile)
	cmd.Dir = "."

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	elapsedTime := time.Since(startTime).Seconds()

	if err != nil {
		return "", false, elapsedTime, true
	}

	// Extract the answer from output (looking for lines like "Part 1: 138")
	for _, line := range strings.Split(stdout.String(), "\n") {
		prefix := fmt.Sprintf("Part %d:", part)
		if strings.Contains(line, prefix) {
			answer := strings.TrimSpace(strings.Split(line, ":")[1])
			return answer, true, elapsedTime, true
		}
	}

	return "", false, elapsedTime, true
}

// getYearsToVerify returns the list of years to verify
func getYearsToVerify(year *int) []int {
	if year != nil {
		return []int{*year}
	}

	// Find all year directories
	entries, err := os.ReadDir(".")
	if err != nil {
		return []int{}
	}

	var years []int
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		yearNum, err := strconv.Atoi(entry.Name())
		if err == nil && yearNum >= 2015 && yearNum <= 2024 {
			years = append(years, yearNum)
		}
	}

	// Sort years
	for i := 0; i < len(years); i++ {
		for j := i + 1; j < len(years); j++ {
			if years[i] > years[j] {
				years[i], years[j] = years[j], years[i]
			}
		}
	}

	return years
}

func main() {
	// Parse command-line arguments
	var yearPos int
	var dayPos int
	var yearFlag int
	var dayFlag int
	var writeMissing bool

	flag.IntVar(&yearFlag, "year", 0, "Year to verify (e.g., 2015)")
	flag.IntVar(&yearFlag, "y", 0, "Year to verify (short form)")
	flag.IntVar(&dayFlag, "day", 0, "Day to verify (1-25)")
	flag.IntVar(&dayFlag, "d", 0, "Day to verify (short form)")
	flag.BoolVar(&writeMissing, "write-missing", false, "Write solution output to missing answer files")
	flag.BoolVar(&writeMissing, "w", false, "Write solution output to missing answer files (short form)")

	flag.Parse()

	// Handle positional arguments
	args := flag.Args()
	if len(args) > 0 {
		if y, err := strconv.Atoi(args[0]); err == nil {
			yearPos = y
		}
	}
	if len(args) > 1 {
		if d, err := strconv.Atoi(args[1]); err == nil {
			dayPos = d
		}
	}

	// Prioritize positional arguments over flags
	var year *int
	if yearPos != 0 {
		year = &yearPos
	} else if yearFlag != 0 {
		year = &yearFlag
	}

	var day *int
	if dayPos != 0 {
		day = &dayPos
	} else if dayFlag != 0 {
		day = &dayFlag
	}

	// Validate day if specified
	if day != nil && (*day < 1 || *day > 25) {
		fmt.Printf("%sError: Day must be between 1 and 25%s\n", colorRed, colorReset)
		os.Exit(1)
	}

	yearsToVerify := getYearsToVerify(year)
	if len(yearsToVerify) == 0 {
		fmt.Printf("%sError: No Go solutions directories found%s\n", colorRed, colorReset)
		os.Exit(1)
	}

	// Determine days to verify
	var daysToVerify []int
	if day != nil {
		daysToVerify = []int{*day}
	} else {
		for i := 1; i <= 25; i++ {
			daysToVerify = append(daysToVerify, i)
		}
	}

	// Verify all years
	allTotalVerified := 0
	allTotalCorrect := 0
	allTotalIncorrect := 0
	allTotalMissing := 0
	allTotalFailed := 0

	for _, yearNum := range yearsToVerify {
		yearDir := strconv.Itoa(yearNum)
		aocDataDir := filepath.Join("..", "aoc-data", yearDir)

		if _, err := os.Stat(yearDir); os.IsNotExist(err) {
			fmt.Printf("%sError: Go solutions directory not found for year %d%s\n", colorRed, yearNum, colorReset)
			fmt.Printf("Expected directory: %s\n", yearDir)
			continue
		}

		if _, err := os.Stat(aocDataDir); os.IsNotExist(err) && !writeMissing {
			fmt.Printf("%sWarning: aoc-data directory not found for year %d%s\n", colorYellow, yearNum, colorReset)
			fmt.Printf("Expected directory: %s\n", aocDataDir)
			fmt.Println("Continuing anyway, but no answers will be verified.")
		}

		if len(yearsToVerify) > 1 {
			fmt.Printf("\n%sVerifying %d Advent of Code Solutions%s\n", colorBlue, yearNum, colorReset)
			fmt.Println(strings.Repeat("=", 60))
		} else {
			fmt.Printf("%sVerifying %d Advent of Code Solutions%s\n", colorBlue, yearNum, colorReset)
			fmt.Println(strings.Repeat("=", 60))
		}

		totalVerified := 0
		totalCorrect := 0
		totalIncorrect := 0
		totalMissing := 0
		totalFailed := 0
		totalWritten := 0

		for _, dayNum := range daysToVerify {
			for part := 1; part <= 2; part++ {
				expected, hasExpected := getExpectedAnswer(yearNum, dayNum, part)

				// Run the solution
				actual, success, elapsedTime, fileExists := runSolution(yearNum, dayNum, part)

				// Case 1: Solution file doesn't exist and answer file doesn't exist
				// -> Skip entirely, no output, no counting
				if !fileExists && !hasExpected {
					continue
				}

				// Case 2: Solution file doesn't exist but answer file exists
				// -> Show MISSING status (solution file is missing)
				if !fileExists && hasExpected {
					fmt.Printf("%s○%s Day %2d Part %d: %sMISSING%s (solution file not found)\n",
						colorYellow, colorReset, dayNum, part, colorYellow, colorReset)
					totalMissing++
					continue
				}

				// Case 3: Solution file exists but failed to run
				if fileExists && (!success || actual == "") {
					emoji := getTimeEmoji(elapsedTime)
					fmt.Printf("%s✗%s Day %2d Part %d: %sFAILED TO RUN%s (%.3fs) %s\n",
						colorRed, colorReset, dayNum, part, colorRed, colorReset, elapsedTime, emoji)
					totalFailed++
					continue
				}

				// Case 4: Solution ran successfully but no expected answer
				// -> Show MISSING status (answer is missing)
				if !hasExpected {
					emoji := getTimeEmoji(elapsedTime)
					statusMsg := fmt.Sprintf("%sMISSING%s", colorYellow, colorReset)

					// Write to answer file if requested
					if writeMissing {
						if writeAnswer(yearNum, dayNum, part, actual) {
							statusMsg = fmt.Sprintf("%sMISSING (wrote: %s)%s", colorYellow, actual, colorReset)
							totalWritten++
						} else {
							statusMsg = fmt.Sprintf("%sMISSING (failed to write)%s", colorYellow, colorReset)
						}
					}

					fmt.Printf("%s○%s Day %2d Part %d: %s (answer: %s, %.3fs) %s\n",
						colorYellow, colorReset, dayNum, part, statusMsg, actual, elapsedTime, emoji)
					totalMissing++
					continue
				}

				// Case 5: Solution ran successfully and has expected answer
				// -> Compare and show CORRECT/INCORRECT
				totalVerified++

				if actual == expected {
					emoji := getTimeEmoji(elapsedTime)
					fmt.Printf("%s✓%s Day %2d Part %d: %sCORRECT%s (answer: %s, %.3fs) %s\n",
						colorGreen, colorReset, dayNum, part, colorGreen, colorReset, actual, elapsedTime, emoji)
					totalCorrect++
				} else {
					emoji := getTimeEmoji(elapsedTime)
					fmt.Printf("%s✗%s Day %2d Part %d: %sINCORRECT%s (expected: %s, got: %s, %.3fs) %s\n",
						colorRed, colorReset, dayNum, part, colorRed, colorReset, expected, actual, elapsedTime, emoji)
					totalIncorrect++
				}
			}
		}

		// Print summary for this year
		fmt.Println()
		fmt.Println(strings.Repeat("=", 60))
		fmt.Printf("%sSummary for %d:%s\n", colorBlue, yearNum, colorReset)
		fmt.Printf("  %sCorrect:%s      %d\n", colorGreen, colorReset, totalCorrect)
		fmt.Printf("  %sIncorrect:%s    %d\n", colorRed, colorReset, totalIncorrect)
		fmt.Printf("  %sFailed:%s       %d\n", colorRed, colorReset, totalFailed)
		fmt.Printf("  %sMissing:%s      %d\n", colorYellow, colorReset, totalMissing)
		if writeMissing && totalWritten > 0 {
			fmt.Printf("  %sWritten:%s      %d\n", colorYellow, colorReset, totalWritten)
		}
		fmt.Printf("  %sTotal Verified:%s %d\n", colorBlue, colorReset, totalVerified)

		allTotalVerified += totalVerified
		allTotalCorrect += totalCorrect
		allTotalIncorrect += totalIncorrect
		allTotalMissing += totalMissing
		allTotalFailed += totalFailed
	}

	// Print overall summary if multiple years
	if len(yearsToVerify) > 1 {
		fmt.Println()
		fmt.Println(strings.Repeat("=", 60))
		fmt.Printf("%sOverall Summary:%s\n", colorBlue, colorReset)
		fmt.Printf("  %sCorrect:%s      %d\n", colorGreen, colorReset, allTotalCorrect)
		fmt.Printf("  %sIncorrect:%s    %d\n", colorRed, colorReset, allTotalIncorrect)
		fmt.Printf("  %sFailed:%s       %d\n", colorRed, colorReset, allTotalFailed)
		fmt.Printf("  %sMissing:%s      %d\n", colorYellow, colorReset, allTotalMissing)
		fmt.Printf("  %sTotal Verified:%s %d\n", colorBlue, colorReset, allTotalVerified)
	}

	if allTotalIncorrect > 0 || allTotalFailed > 0 {
		os.Exit(1)
	} else {
		if allTotalVerified > 0 {
			fmt.Printf("\n%sAll verified solutions are correct!%s\n", colorGreen, colorReset)
		} else {
			fmt.Printf("\n%sNo solutions were verified.%s\n", colorYellow, colorReset)
		}
	}
}

File: javascript/verify_solutions.js
#!/usr/bin/env node

/**
 * Verify Advent of Code solutions against known correct answers.
 *
 * This script runs each solution file for a specified year and compares the output
 * against the answer files stored in aoc-data.
 *
 * Usage:
 *     node verify_solutions.js [YEAR] [DAY] [OPTIONS]
 *     npm run verify [-- YEAR] [DAY] [OPTIONS]
 *
 * Positional Arguments:
 *     YEAR                      The year to verify (e.g., 2015, 2016, 2017)
 *     DAY                       The day to verify (1-25)
 *
 * Options:
 *     --year YEAR, -y YEAR      The year to verify (alternative to positional)
 *     --day DAY, -d DAY         The day to verify (alternative to positional)
 *     --write-missing, -w       Write solution output to missing answer files
 *
 * Examples:
 *     npm run verify                              # Verify all years
 *     npm run verify -- 2015                      # Verify year 2015
 *     npm run verify -- 2015 20                   # Verify year 2015, day 20
 *     npm run verify -- --year 2015 --day 20 --write-missing
 *
 * If no year is specified, all years will be verified.
 * If no day is specified, all days will be verified.
 */

import { readFileSync, readdirSync, existsSync, mkdirSync, writeFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { spawn } from 'child_process';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Colors for terminal output
const GREEN = '\x1b[92m';
const RED = '\x1b[91m';
const YELLOW = '\x1b[93m';
const BLUE = '\x1b[94m';
const RESET = '\x1b[0m';

/**
 * Get an emoji based on the execution time.
 *
 * @param {number} elapsedTime - Execution time in seconds
 * @returns {string} Emoji string representing the speed
 */
function getTimeEmoji(elapsedTime) {
  if (elapsedTime < 1.0) return '⚡';
  if (elapsedTime < 3.0) return '🚀';
  if (elapsedTime < 10.0) return '▶️';
  if (elapsedTime < 30.0) return '🐢';
  return '🐌';
}

/**
 * Read the expected answer from aoc-data directory.
 *
 * @param {number} day - Day number (1-25)
 * @param {number} part - Part number (1 or 2)
 * @param {number} year - Year (e.g., 2015, 2016, 2017)
 * @param {string} baseDir - Base directory of the javascript solutions
 * @returns {string|null} Expected answer as string, or null if file doesn't exist or is empty
 */
function getExpectedAnswer(day, part, year, baseDir) {
  // Path from javascript/ dir to aoc-data is ../aoc-data
  const answerFile = join(baseDir, '..', 'aoc-data', String(year), String(day), `solution-${part}`);

  if (!existsSync(answerFile)) {
    return null;
  }

  try {
    const content = readFileSync(answerFile, 'utf-8').trim();
    // Return null if empty or just '0'
    if (content && content !== '0') {
      return content;
    }
    return null;
  } catch (error) {
    return null;
  }
}

/**
 * Write an answer to the aoc-data directory.
 *
 * @param {number} day - Day number (1-25)
 * @param {number} part - Part number (1 or 2)
 * @param {number} year - Year (e.g., 2015, 2016, 2017)
 * @param {string} answer - The answer to write
 * @param {string} baseDir - Base directory of the javascript solutions
 * @returns {boolean} True if successfully written, false otherwise
 */
function writeAnswer(day, part, year, answer, baseDir) {
  // Path from javascript/ dir to aoc-data is ../aoc-data
  const answerDir = join(baseDir, '..', 'aoc-data', String(year), String(day));
  const answerFile = join(answerDir, `solution-${part}`);

  try {
    // Create directory if it doesn't exist
    mkdirSync(answerDir, { recursive: true });

    // Write the answer
    writeFileSync(answerFile, answer.trim() + '\n', 'utf-8');
    return true;
  } catch (error) {
    console.error(`${RED}Error writing answer: ${error.message}${RESET}`);
    return false;
  }
}

/**
 * Run a solution file and extract the answer.
 *
 * @param {number} day - Day number (1-25)
 * @param {number} part - Part number (1 or 2)
 * @param {number} year - Year (e.g., 2015, 2016, 2017)
 * @param {string} baseDir - Base directory of the javascript solutions
 * @returns {Promise<{answer: string|null, success: boolean, elapsedTime: number, fileExists: boolean}>}
 */
async function runSolution(day, part, year, baseDir) {
  const dayPadded = String(day).padStart(2, '0');
  const solutionFile = join(baseDir, String(year), dayPadded, `solution${part}.js`);

  if (!existsSync(solutionFile)) {
    return { answer: null, success: false, elapsedTime: 0.0, fileExists: false };
  }

  return new Promise((resolve) => {
    const startTime = Date.now();
    const child = spawn('node', [solutionFile], {
      cwd: dirname(solutionFile),
      timeout: 30000,
    });

    let stdout = '';
    let stderr = '';

    child.stdout.on('data', (data) => {
      stdout += data.toString();
    });

    child.stderr.on('data', (data) => {
      stderr += data.toString();
    });

    child.on('close', (code) => {
      const elapsedTime = (Date.now() - startTime) / 1000;

      if (code !== 0) {
        resolve({ answer: null, success: false, elapsedTime, fileExists: true });
        return;
      }

      // Extract the answer from output (looking for lines like "Part 1: 138")
      const lines = stdout.split('\n');
      for (const line of lines) {
        if (line.includes(`Part ${part}:`)) {
          const answer = line.split(':').slice(-1)[0].trim();
          resolve({ answer, success: true, elapsedTime, fileExists: true });
          return;
        }
      }

      resolve({ answer: null, success: false, elapsedTime, fileExists: true });
    });

    child.on('error', () => {
      const elapsedTime = (Date.now() - startTime) / 1000;
      resolve({ answer: null, success: false, elapsedTime, fileExists: true });
    });
  });
}

/**
 * Get the list of years to verify.
 *
 * @param {string} baseDir - Base directory of the javascript solutions
 * @param {number|null} year - Specific year to verify, or null for all years
 * @returns {number[]} List of years to verify
 */
function getYearsToVerify(baseDir, year) {
  if (year !== null) {
    return [year];
  }

  // Find all years with javascript solutions
  if (!existsSync(baseDir)) {
    return [];
  }

  const years = [];
  const items = readdirSync(baseDir, { withFileTypes: true });
  for (const item of items) {
    if (item.isDirectory() && /^\d{4}$/.test(item.name)) {
      years.push(parseInt(item.name, 10));
    }
  }

  return years.sort((a, b) => a - b);
}

/**
 * Parse command line arguments.
 *
 * @param {string[]} args - Command line arguments
 * @returns {{year: number|null, day: number|null, writeMissing: boolean}}
 */
function parseArgs(args) {
  let year = null;
  let day = null;
  let writeMissing = false;

  // Check for positional arguments first
  const positionalArgs = args.filter((arg) => !arg.startsWith('-'));
  if (positionalArgs.length > 0) {
    const yearArg = parseInt(positionalArgs[0], 10);
    if (!isNaN(yearArg)) {
      year = yearArg;
    }
  }
  if (positionalArgs.length > 1) {
    const dayArg = parseInt(positionalArgs[1], 10);
    if (!isNaN(dayArg)) {
      day = dayArg;
    }
  }

  // Check for flag arguments (override positional)
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if ((arg === '--year' || arg === '-y') && i + 1 < args.length) {
      const yearArg = parseInt(args[i + 1], 10);
      if (!isNaN(yearArg)) {
        year = yearArg;
      }
      i++;
    } else if ((arg === '--day' || arg === '-d') && i + 1 < args.length) {
      const dayArg = parseInt(args[i + 1], 10);
      if (!isNaN(dayArg)) {
        day = dayArg;
      }
      i++;
    } else if (arg === '--write-missing' || arg === '-w') {
      writeMissing = true;
    }
  }

  return { year, day, writeMissing };
}

/**
 * Main verification function.
 */
async function main() {
  const args = process.argv.slice(2);
  const { year, day, writeMissing } = parseArgs(args);

  const baseDir = __dirname;
  const yearsToVerify = getYearsToVerify(baseDir, year);

  if (yearsToVerify.length === 0) {
    console.error(`${RED}Error: No JavaScript solutions directories found${RESET}`);
    process.exit(1);
  }

  // Determine days to verify
  let daysToVerify;
  if (day !== null) {
    if (day < 1 || day > 25) {
      console.error(`${RED}Error: Day must be between 1 and 25${RESET}`);
      process.exit(1);
    }
    daysToVerify = [day];
  } else {
    daysToVerify = Array.from({ length: 25 }, (_, i) => i + 1);
  }

  // Verify all years
  let allTotalVerified = 0;
  let allTotalCorrect = 0;
  let allTotalIncorrect = 0;
  let allTotalMissing = 0;
  let allTotalFailed = 0;

  for (const year of yearsToVerify) {
    const jsYearDir = join(baseDir, String(year));
    const aocDataDir = join(baseDir, '..', 'aoc-data', String(year));

    if (!existsSync(jsYearDir)) {
      console.error(`${RED}Error: JavaScript solutions directory not found for year ${year}${RESET}`);
      console.error(`Expected directory: ${jsYearDir}`);
      continue;
    }

    if (!existsSync(aocDataDir) && !writeMissing) {
      console.warn(`${YELLOW}Warning: aoc-data directory not found for year ${year}${RESET}`);
      console.warn(`Expected directory: ${aocDataDir}`);
      console.warn('Continuing anyway, but no answers will be verified.');
    }

    if (yearsToVerify.length > 1) {
      console.log(`\n${BLUE}Verifying ${year} Advent of Code Solutions${RESET}`);
      console.log('='.repeat(60));
    } else {
      console.log(`${BLUE}Verifying ${year} Advent of Code Solutions${RESET}`);
      console.log('='.repeat(60));
    }

    let totalVerified = 0;
    let totalCorrect = 0;
    let totalIncorrect = 0;
    let totalMissing = 0;
    let totalFailed = 0;
    let totalWritten = 0;

    for (const day of daysToVerify) {
      for (const part of [1, 2]) {
        const expected = getExpectedAnswer(day, part, year, baseDir);

        // Run the solution
        const { answer: actual, success, elapsedTime, fileExists } = await runSolution(day, part, year, baseDir);

        // Case 1: Solution file doesn't exist and answer file doesn't exist
        // -> Skip entirely, no output, no counting
        if (!fileExists && expected === null) {
          continue;
        }

        // Case 2: Solution file doesn't exist but answer file exists
        // -> Show MISSING status (solution file is missing)
        if (!fileExists && expected !== null) {
          console.log(
            `${YELLOW}○${RESET} Day ${String(day).padStart(2)} Part ${part}: ${YELLOW}MISSING${RESET} (solution file not found)`
          );
          totalMissing++;
          continue;
        }

        // Case 3: Solution file exists but failed to run
        if (fileExists && (!success || actual === null)) {
          const emoji = getTimeEmoji(elapsedTime);
          console.log(
            `${RED}✗${RESET} Day ${String(day).padStart(2)} Part ${part}: ${RED}FAILED TO RUN${RESET} (${elapsedTime.toFixed(3)}s) ${emoji}`
          );
          totalFailed++;
          continue;
        }

        // Case 4: Solution ran successfully but no expected answer
        // -> Show MISSING status (answer is missing)
        if (expected === null) {
          const emoji = getTimeEmoji(elapsedTime);
          let statusMsg = `${YELLOW}MISSING${RESET}`;

          // Write to answer file if requested
          if (writeMissing) {
            if (writeAnswer(day, part, year, actual, baseDir)) {
              statusMsg = `${YELLOW}MISSING (wrote: ${actual})${RESET}`;
              totalWritten++;
            } else {
              statusMsg = `${YELLOW}MISSING (failed to write)${RESET}`;
            }
          }

          console.log(
            `${YELLOW}○${RESET} Day ${String(day).padStart(2)} Part ${part}: ${statusMsg} (answer: ${actual}, ${elapsedTime.toFixed(3)}s) ${emoji}`
          );
          totalMissing++;
          continue;
        }

        // Case 5: Solution ran successfully and has expected answer
        // -> Compare and show CORRECT/INCORRECT
        totalVerified++;

        if (actual === expected) {
          const emoji = getTimeEmoji(elapsedTime);
          console.log(
            `${GREEN}✓${RESET} Day ${String(day).padStart(2)} Part ${part}: ${GREEN}CORRECT${RESET} (answer: ${actual}, ${elapsedTime.toFixed(3)}s) ${emoji}`
          );
          totalCorrect++;
        } else {
          const emoji = getTimeEmoji(elapsedTime);
          console.log(
            `${RED}✗${RESET} Day ${String(day).padStart(2)} Part ${part}: ${RED}INCORRECT${RESET} (expected: ${expected}, got: ${actual}, ${elapsedTime.toFixed(3)}s) ${emoji}`
          );
          totalIncorrect++;
        }
      }
    }

    // Print summary for this year
    console.log('\n' + '='.repeat(60));
    console.log(`${BLUE}Summary for ${year}:${RESET}`);
    console.log(`  ${GREEN}Correct:${RESET}      ${totalCorrect}`);
    console.log(`  ${RED}Incorrect:${RESET}    ${totalIncorrect}`);
    console.log(`  ${RED}Failed:${RESET}       ${totalFailed}`);
    console.log(`  ${YELLOW}Missing:${RESET}      ${totalMissing}`);
    if (writeMissing && totalWritten > 0) {
      console.log(`  ${YELLOW}Written:${RESET}      ${totalWritten}`);
    }
    console.log(`  ${BLUE}Total Verified:${RESET} ${totalVerified}`);

    allTotalVerified += totalVerified;
    allTotalCorrect += totalCorrect;
    allTotalIncorrect += totalIncorrect;
    allTotalMissing += totalMissing;
    allTotalFailed += totalFailed;
  }

  // Print overall summary if multiple years
  if (yearsToVerify.length > 1) {
    console.log('\n' + '='.repeat(60));
    console.log(`${BLUE}Overall Summary:${RESET}`);
    console.log(`  ${GREEN}Correct:${RESET}      ${allTotalCorrect}`);
    console.log(`  ${RED}Incorrect:${RESET}    ${allTotalIncorrect}`);
    console.log(`  ${RED}Failed:${RESET}       ${allTotalFailed}`);
    console.log(`  ${YELLOW}Missing:${RESET}      ${allTotalMissing}`);
    console.log(`  ${BLUE}Total Verified:${RESET} ${allTotalVerified}`);
  }

  if (allTotalIncorrect > 0 || allTotalFailed > 0) {
    process.exit(1);
  } else {
    if (allTotalVerified > 0) {
      console.log(`\n${GREEN}All verified solutions are correct!${RESET}`);
    } else {
      console.log(`\n${YELLOW}No solutions were verified.${RESET}`);
    }
  }
}

main().catch((error) => {
  console.error(`${RED}Fatal error: ${error.message}${RESET}`);
  process.exit(1);
});
